packages/navigator/src/browser/open-editors-widget/navigator-deleted-editor-decorator.ts

```diff
@@ -0,0 +1,77 @@
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, postConstruct } from '@theia/core/shared/inversify';
+import { DepthFirstTreeIterator, NavigatableWidget, Tree, TreeDecoration, TreeDecorator } from '@theia/core/lib/browser';
+import { FileSystemFrontendContribution } from '@theia/filesystem/lib/browser/filesystem-frontend-contribution';
+import { Emitter } from '@theia/core';
+import { OpenEditorNode } from './navigator-open-editors-tree-model';
+import { FileChangeType } from '@theia/filesystem/lib/browser';
+
+@injectable()
+export class NavigatorDeletedEditorDecorator implements TreeDecorator {
+
+    @inject(FileSystemFrontendContribution)
+    protected readonly fileSystemContribution: FileSystemFrontendContribution;
+
+    readonly id = 'theia-deleted-editor-decorator';
+    protected readonly onDidChangeDecorationsEmitter = new Emitter();
+    readonly onDidChangeDecorations = this.onDidChangeDecorationsEmitter.event;
+    protected deletedEditors = new Set<NavigatableWidget>();
+
+    @postConstruct()
+    init(): void {
+        this.fileSystemContribution.onDidChangeEditorFile(({ editor, type }) => {
+            if (type === FileChangeType.DELETED) {
+                this.deletedEditors.add(editor);
+            } else if (type === FileChangeType.ADDED) {
+                this.deletedEditors.delete(editor);
+            }
+            this.fireDidChangeDecorations((tree: Tree) => this.collectDecorators(tree));
+        });
+    }
+
+    decorations(tree: Tree): Map<string, TreeDecoration.Data> {
+        return this.collectDecorators(tree);
+    }
+
+    protected collectDecorators(tree: Tree): Map<string, TreeDecoration.Data> {
+        const deletedEditorsCopy = new Set(this.deletedEditors);
+        const result = new Map<string, TreeDecoration.Data>();
+        if (tree.root === undefined) {
+            return result;
+        }
+        for (const node of new DepthFirstTreeIterator(tree.root)) {
+            if (OpenEditorNode.is(node) && NavigatableWidget.is(node.widget)) {
+                if (this.deletedEditors.has(node.widget)) {
+                    const deletedDecoration: TreeDecoration.Data = {
+                        fontData: {
+                            style: 'line-through',
+                        }
+                    };
+                    deletedEditorsCopy.delete(node.widget);
+                    result.set(node.id, deletedDecoration);
+                }
+            }
+        }
+        deletedEditorsCopy.forEach(remainingEditor => this.deletedEditors.delete(remainingEditor));
```

Comment:
This code could behave incorrectly if this decorator were used in a (hypothetical) 'Open TypeScript Editors' widget. Then any deleted files that were not TypeScript would be removed from the deleted tracker when the decorator was applied to that tree and be unavailable to (non-hypothetical) general Open Editors widgets. It might be better to listen to editor closures and remove things from the tracking set then?

Reply:
Good suggestion, I've updated the tracking system for the deleted editors to rebuild the `this.deletedEditors` in response to onDidAddWidget